Segger RTT: support CONFIG_SEGGER_RTT_INIT_MODE and optimizations

giansta opened the pull request:
* Different initialization modes of Control Block
 * Safer INIT() like in SystemView V352a
 * Made _aTerminalId const like in SystemView V352a

Fixes #53544.

giansta commented:
Expected to fix [Issue #53544](https://github.com/zephyrproject-rtos/zephyr/issues/53544) with [PR  #53569](https://github.com/zephyrproject-rtos/zephyr/pull/53569) in Zephyr repo

nordic-krch commented on SEGGER/SEGGER_RTT.c:
```diff
@@ -167,6 +167,26 @@ Additional information:
   #endif
 #endif
 
+#ifndef   SEGGER_RTT_MEMCMP
```
is this a new segger version or your modifications?
I'm afraid that those changes may get lost if someone updates with new version of segger_rtt.

giansta replied:
It's my modifications. I was assuming that Segger people follow this repo and approve these changes so that they become they also take them.
If not, what do you suggest? Contact them through their official website?

giansta replied:
When I look in previous commits, they don't seem to be done by Segger, like my one. Are there some additional rules to coordinate with Segger?

nordic-krch replied:
I'm not sure how to push it to segger.

giansta replied:
Maybe @sigvartmh or @carlescufi can help?

giansta replied:
> Segger is not part of the zephyrproject-rtos Git-Repo.
> However, we will review the changes and extend the official RTT sources if we
> come to the conclusion that it is a good addition to the current RTT state.
> I will keep you posted.

I got this reply after opening a ticket in Segger support.

giansta replied:
After opening a ticket in support.segger.com, they proposed the update V7.86h that gives equivalent changes and improved initialization. I tried to leave the `SEGGER_RTT.c` untouched and follow their suggestion to not call at all `SEGGER_RTT_Init()` in the chained application, but, even if working, I saw that the APIs using use `INIT()` aren't used in Zephyr log backend. So I added a change only in `SEGGER_RTT_Init()`, using the configuration I'm aware of that distinguish the cases when we want to to a forced init of CB or just check if it's already done.

nordic-krch commented on SEGGER/SEGGER_RTT.c:
```diff
@@ -167,6 +167,26 @@ Additional information:
+
+#ifndef   SEGGER_RTT_MEMSET
+  #ifdef  MEMSET
+    #define SEGGER_RTT_MEMSET(pDest, FillChar, NumBytes)  MEMCPY((pDest), (FillChar), (NumBytes))
```
copy paste error. `MEMCPY` -> `MEMSET`

giansta replied:
Thanks for spotting this, fixed.

MarcianoPreciado commented:
Can we get this reviewed? This is causing me issues also

carlescufi commented:
@sigvartmh please review this as part of this change
@giansta please rebase since we've updated to RTT 3.40

MarekPieta commented on SEGGER/SEGGER_RTT.c:
```diff
@@ -1876,7 +1899,20 @@ int SEGGER_RTT_SetFlagsDownBuffer(unsigned BufferIndex, unsigned Flags) {
+   * Segger code using _DoInit() is used. For applications started by
+   * bootloader, Segger recommends not calling this function, as all the
+   * recommended APIs use INIT(). Unfortunately Zephyr log backend uses
+   * SEGGER_RTT_WriteSkipNoLock() that doesn't call INT(), so we need
```
typo: `INIT`

MarekPieta replied:
Maybe we could add a dedicated Kconfig option to let user decide if Zephyr application should actually force overwrite of the RTT structures during initialization. The current approach cannot support e.g. other bootloaders.
That would also allow to preserve RTT content during e.g. software reboot that does not clear RAM content.

MarekPieta replied:
I also wonder if it wouldn't be reasonable to check whole `acID` right before initialization (instead of checking just the first byte) to ensure no false positives in case RAM content is not zero-initialized (with current implementation chance of false positive when checking if structure is initialized is quite high - `~1/255`).
Actually after we improve the check, we may event consider to always call `INIT` here

MarekPieta replied:
If we are going to rely on bootloader to initialize the RTT structures, we need to ensure backwards compatibility of the initialization operation (many embedded applications rely on immutable bootloader that cannot be updated once flashed).

giansta replied:
Thanks for your suggestions. Unfortunately I was too busy and I'll be out now for a couple of weeks, but afterwards I'll try to follow them.
The Kconfig option is a good idea, I was not very satisfied with my `#ifdefs` that are less general. I just have to figure out where such a Kconfig can be placed.
I also agree on the whole `acID` check, the `INIT()` macro can be used like it is only for other functions than `SEGGER_RTT_Init()`. I tried not to touch the SEGGER module, then trying with minimal change, but this is indeed reasonable.
I'll even rebase to RTT 3.40.

giansta replied:
With latest release I tried to take account of all above suggestions:
 * Rebased to SystemView V340
 * Managed dedicated CONFIG_SEGGER_RTT_INIT_MODE to have different init behaviours

The second change gives more flexibility: application level configs can force initialization of control block without condition or have a complete or partial check on acID to have a contitional init. Yet this requires some more changes in Segger code.

The SystemView V340 code misses some minor changes (safer INIT(), const _aTerminalId) tht are provided by latest SystemView V352a: I've added them, as in previous version.

The CONFIG_SEGGER_RTT_INIT_MODE is defined in [PR #53569](https://github.com/zephyrproject-rtos/zephyr/pull/53569) in Zephyr repo.

sigvartmh replied:
Thanks for all your changes looks good :) I've done similar modifications but haven't pushed changes to Zephyr due to having it interop with other parts of downstream Nordic Fork, which I couldn't figure out how to handle in a nice way. However having this upstream would definitely be a good addition.

giansta committed:
Segger RTT: support CONFIG_SEGGER_RTT_INIT_MODE and optimizations

Support Zephyr "RTT Initialization mode" config choice, allowing
Control Block initialization to be retained from another program
(e.g. bootloader).
It includes conditional CB init with full check on Control Block ID or
partial one that checks only the first byte, yet improving its logic.
Minor optimizations with static const added.

Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>

giansta committed:
segger: config: support RTT section custom definition in Zephyr DTS

Allow to define the name of custom RTT section both when it's configured
to be placed at RAM start and when its address is defined by a memory
region in DTS.

Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>

henrikbrixandersen reviewed (commented):
I generally think this looks good, but it could benefit from either a more elaborate Git commit message describing the changes in detail, or more code comments - or both.

giansta commented:
> I generally think this looks good, but it could benefit from either a more elaborate Git commit message describing the changes in detail, or more code comments - or both.

I tried to have the commit messages clearer.
With regard to comments, I believe that most code changes are taken from Segger code and they should be quite clear.
The key point is in `SEGGER_RTT_Init()`, where I wrote a comment explaining the reason of changes.
If you feel that something else could be enhanced, please let me know.

henrikbrixandersen commented on Config/SEGGER_RTT_Conf.h:
```diff
@@ -98,7 +98,7 @@ Revision: $Rev: 24316 $
 #elif defined(CONFIG_SEGGER_RTT_SECTION_CCM)
 #define SEGGER_RTT_SECTION                        ".ccm_data"
-#elif defined(CONFIG_SEGGER_RTT_SECTION_CUSTOM)
+#elif defined(CONFIG_SEGGER_RTT_SECTION_CUSTOM) || defined(CONFIG_SEGGER_RTT_SECTION_CUSTOM_DTS_REGION)
```
Where is `CONFIG_SEGGER_RTT_SECTION_CUSTOM_DTS_REGION` declared?

sigvartmh replied:
Ref see last comment https://github.com/zephyrproject-rtos/zephyr/pull/53569/files

giansta replied:
Correct, this PR depends on [PR #53569](https://github.com/zephyrproject-rtos/zephyr/pull/53569): there the `choice SEGGER_RTT_SECTION` in `modules/segger/Kconfig` has been extended,  `SEGGER_RTT_SECTION_CUSTOM_NAME` and `choice SEGGER_RTT_INIT_MODE` have been added.

henrikbrixandersen commented on SEGGER/SEGGER_RTT.c:
```diff
@@ -1892,6 +1893,28 @@ int SEGGER_RTT_SetFlagsDownBuffer(unsigned BufferIndex, unsigned Flags) {
   return r;
 }
 
+#if defined(CONFIG_SEGGER_RTT_INIT_MODE_STRONG_CHECK)
```
Where is `CONFIG_SEGGER_RTT_INIT_MODE_STRONG_CHECK` declared?

sigvartmh replied:
Ref. see last comment https://github.com/zephyrproject-rtos/zephyr/pull/53569/files

giansta replied:
same as above

henrikbrixandersen commented on SEGGER/SEGGER_RTT.c:
```diff
@@ -1902,7 +1925,22 @@ int SEGGER_RTT_SetFlagsDownBuffer(unsigned BufferIndex, unsigned Flags) {
+   */
+#if defined(CONFIG_SEGGER_RTT_INIT_MODE_STRONG_CHECK)
+  SEGGER_RTT_StrongCheckInit();
+#elif defined(CONFIG_SEGGER_RTT_INIT_MODE_WEAK_CHECK)
```
Where is `CONFIG_SEGGER_RTT_INIT_MODE_WEAK_CHECK` declared?

sigvartmh replied:
I expected them to come from here https://github.com/zephyrproject-rtos/zephyr/pull/53569/files

 Do we need to add another layer. I think that's only reasonable if this change is expected to be merged into Seggers actual source code.
 
 Please correct me if I'm wrong.

giansta replied:
same as above

cfriedt commented on SEGGER/SEGGER_RTT.c:
```diff
@@ -1892,6 +1893,28 @@ int SEGGER_RTT_SetFlagsDownBuffer(unsigned BufferIndex, unsigned Flags) {
+  volatile SEGGER_RTT_CB* p;
+  p = (volatile SEGGER_RTT_CB*)((char*)&_SEGGER_RTT + SEGGER_RTT_UNCACHED_OFF);
+  unsigned i;
+  bool DoInit = false;
```
I needed to include `<stdbool.h>` at the top of this file to avoid this error when building against `v3.6-branch`
```shell
SEGGER/SEGGER_RTT.c:1906:3: error: unknown type name 'bool'
 1906 |   bool DoInit = false;
```

giansta replied:
Thanks @cfriedt, for signalling this issue. It probably depends on the toolchain, as I didn't have it.
Yet, looking at Seger code, they seem not to use boolean, so I changed it to int without using false/true values.

cfriedt commented on SEGGER/SEGGER_RTT.c:
```diff
@@ -1892,6 +1893,28 @@ int SEGGER_RTT_SetFlagsDownBuffer(unsigned BufferIndex, unsigned Flags) {
+  RTT__DMB();                       // Force order of memory accesses for cores that may perform out-of-order memory accesses
+  for (i = 0; i < sizeof(_aInitStr) - 1; ++i) {
+    if (p->acID[i] != _aInitStr[sizeof(_aInitStr) - 2 - i]) { // Skip terminating \0 at the end of the array
+    	DoInit = true;
```
Similar error / need for `<stdbool.h>`
```shell
SEGGER/SEGGER_RTT.c:1913:16: error: 'true' undeclared (first use in this function)
 1913 |       DoInit = true;
```

giansta replied:
As above

henrikbrixandersen reviewed (approved):
LGTM

brycewilkins commented:
@giansta I tried to get the Zephyr repo `west.yml` updated to have this pull request merged (https://github.com/zephyrproject-rtos/zephyr/pull/81425), but this PR needs to be rebased. Could you do that? Would love to have this in!

cfriedt commented:
We should really have a rule that source code should be checked in with UNIX line endings. The sources in this repository are somewhat difficult to work with because they have DOS line endings.

brycewilkins commented:
> We should really have a rule that source code should be checked in with UNIX line endings. The sources in this repository are somewhat difficult to work with because they have DOS line endings.

Agree; let's address that in a new PR; I can do it.

brycewilkins commented:
@giansta Thanks for rebase!

@cfriedt @nordic-krch @nordicjm Could this be reviewed so we can get it over the line? 
- The Zephyr repo PR is now passing all tests after the rebase: https://github.com/zephyrproject-rtos/zephyr/pull/81425

brycewilkins reviewed (approved):
lgtm! This will be great to have merged!
